### PR TITLE
Enhance ThingResource in order to create Things with the dynamically provided Channels

### PR DESCRIPTION
**Pull request by teichsta**
While creating Things (through REST) their Channels are adding according to the given `channel-configuration` of their ESH-INF/configuration files. This PR add the possibility to provide the channels to be created over REST.

This add the possibility to create Things with variable amount of channels.

### CODE REVIEW

**Comment by kaikreuzer** [comment]
I am not completely clear on the context. Is this really only an enhancement or does this actually address a bug (which has already been reported/discussed anywhere)?
From the code, it seems that channels on a thing passed through REST are currently ignored and this change makes them being considered. Is this the correct interpretation?

**Comment by teichsta** [comment]
> I am not completely clear on the context. Is this really only an enhancement or does this actually address a bug (which has already been reported/discussed anywhere)?

this PR is really about an enhancement which is not discussed anywhere.

> From the code, it seems that channels on a thing passed through REST are currently ignored and this change makes them being considered. Is this the correct interpretation?

Yes, absolutely! This PR is the ESH enhancement being necessary to contribute the GenericKnxHandler which will follow in the coming days.

**Comment by kaikreuzer** [comment]
>>  Is this the correct interpretation?

> Yes, absolutely! 

Well, then I would consider it a bug fix, not an enhancement :-)
Could you please also add a small unit test, at least for the new method of the ManagedThingProvider?

File: bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/thing/ThingResource.java [hunk 1]
```diff
@@ -120,16 +124,32 @@ public Response create(@HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) @ApiParam(value
...
         Thing thing = thingRegistry.get(thingUIDObject);
 
         // does the Thing already exist?
-        if (null == thing) {
-            // if not, create new Thing
+        if (thing == null) {
+        	
+            // if not, create new Channels and add them to a new Thing
+            List<Channel> channels = new ArrayList<Channel>();
+            for (ChannelDTO channelBean : thingBean.channels) {
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/thing/ThingResource.java]: Is is guaranteed that `thingBean.channels` is not null?

**Comment by teichsta** [comment]
> Could you please also add a small unit test, at least for the new method of the ManagedThingProvider?
> 

i’ll have a look at this at the weekend latest …

File: bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/thing/ThingResource.java [hunk 2]
```diff
@@ -120,16 +124,32 @@ public Response create(@HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) @ApiParam(value
...
+            	
+            	String channelId = channelBean.id;
+            	if (channelId == null) {
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/thing/ThingResource.java]: Is this at all a valid situation? I would assume that a client has to provide an id here.
Note: For rule modules within a rule, we expect the same, so requiring an id here would be in line with that.

**Comment by kaikreuzer** [comment]
Looking a bit closer at the (existing, old) code, I feel that the `ThingResource.create` might need to be completely revisited. Besides swallowing the channels, it does the same with properties, doesn't it?
So I would suggest, that the ThingDTOMapper gets enhanced to also map ThingDTOs to Things and that this is then called by the create method and the result is fed into the ThingRegistry (not the ManagedThingProvider anymore!)
The very same would have to be done in the `ThingResource.update` method, as this swallows all newly added (or removed!) channels and properties as well.

**Comment by teichsta** [comment]
updated the PR (with the changes before your latest comment)

> Besides swallowing the channels, it does the same with properties, doesn't it?

yes, that seems so … however with respect to the changes of the KNX binding i am going to contribute the next days it would be great to have this merge first and then start the bigger change. Would you agree?

**Comment by kgoderis** [comment]
@teichsta FYI I made some changes this week to the proposer KNX pr to adjust to the latest ESH APIs. No extension just render it compatible

**Comment by teichsta** [comment]
Thanks, Karel … will have a look the next week.

**Comment by kaikreuzer** [comment]
@teichsta As you know, temporary solutions last the longest. Since the code of this PR would not even stay once it is implemented as suggested, I don't think that merging like this makes sense - especially as it seems that we need the correct fix right away, see https://github.com/eclipse/smarthome/issues/1103. I don't think that it is much work to do it right straight away.

I wonder how your KNX binding enhancements could possibly depend on changes in the REST API? There imho should be no dependency whatsoever... In any case, feel free to create the KNX PR and add a note that it must not be merged before a certain feature is available in ESH.

**Comment by kaikreuzer** [comment]
@teichsta Any update?

**Comment by teichsta** [comment]
thanks Kai für providing #1286 meanwhile. I am closing this PR since yours seem to cover the whole issue, right?

**Comment by kaikreuzer** [comment]
Right - I hope it works as expected for you.

**Pull request closed without merging**